db/python/tables/meta_table.py

```diff
@@ -0,0 +1,144 @@
+EXTERNAL_ORG_SENTINEL = '__PRIMARY__EXTERNAL__ORG__'
+
+
+class MetaTable(DbBase):
+    """
+    Generate a flat table with metadata for all entity tables
+    """
+
+    def external_id_query(self, table_alias: str):
+        """
+        Returns the query partial to aggregate external ids in to a json object
+        including replacing the primary external org with a sentinel value. This is
+        required because duckdb doesn't support column names which are an empty string.
+        """
+        return f"""
+            JSON_OBJECTAGG(
+                CASE
+                    WHEN {table_alias}.name = :primary_external_org
+                    THEN :primary_external_org_sentinel
+                    ELSE {table_alias}.name
+                END,
+                {table_alias}.external_id
+            ) AS external_ids
+        """
+
+    async def entity_meta_table(
+        self,
+        project: int,
+        query: str,
+        row_getter: Callable[[Record], dict[str, Any]],
+        has_external_ids: bool,
+        has_meta: bool,
+    ):
+        """
+        Return a flat tabular parquet file for the provied query. Optionally include
+        metadata and external_ids columns. This returns a BytesIO object with the
+        parquet file.
+        """
+
+        rows = await self.connection.fetch_all(
+            query,
+            {
+                'project': project,
+                'primary_external_org': PRIMARY_EXTERNAL_ORG,
+                'primary_external_org_sentinel': EXTERNAL_ORG_SENTINEL,
+            },
+        )
+
+        if len(rows) == 0:
+            return None
+
+        duck = duckdb.connect()
+        meta_rows_str = StringIO()
+        external_id_rows_str = StringIO()
+        main_rows: list[dict[str, Any]] = []
+
+        for row in rows:
+            if has_meta:
+                meta_rows_str.write((row['meta'] or 'null') + '\n')
+
+            if has_external_ids:
+                external_id_rows_str.write(
+                    (row['external_ids'] or f'{{"{EXTERNAL_ORG_SENTINEL}": null}}')
+                    + '\n'
+                )
+
+            main_rows.append(row_getter(row))
+
+        duck.register('main_table', pa.Table.from_pylist(main_rows))
+
+        # Seek str buffers back to beginning
+        meta_rows_str.seek(0)
+        external_id_rows_str.seek(0)
+
+        # parse the json for the meta table
+        meta_columns = ''
+        meta_join = ''
+        if has_meta:
+            duck.register(
+                'meta_table',
+                duck.read_json(
+                    meta_rows_str,
+                    map_inference_threshold=-1,
+                    format='newline_delimited',
+                ).arrow(),
+            )
+            # Prefix all meta columns with `meta_` to avoid clashes with the main table
+            meta_columns = """
+                ,COLUMNS(mt.*) as 'meta_\\0'
+            """
+            meta_join = 'POSITIONAL JOIN meta_table mt'
```

Comment:
Can't say I've ever seen a `POSITIONAL JOIN` before. Interesting, how does it work. More just curious than anything.

Reply:
It is basically like `zip` in python, or a horizontal concat in dataframes. It take two tables and joins them based on the row number. Here's the docs for it https://duckdb.org/docs/sql/query_syntax/from.html#positional-joins